eloquentjs/20_1_search_tool/src/nodeGREP.ts

```diff
@@ -0,0 +1,54 @@
+import * as  fs from "fs"
+
+
+let regeExpString = process.argv[2];
+let filesAndFolders = process.argv.slice(3);
+
+if (!regeExpString || filesAndFolders.length === 0) {
+    throw "Missing arguments"
+}
+
+console.log(nodeGREP(regeExpString, filesAndFolders));
+
+/**
+ * Searches provided files, folders, and subfolders for file content that matches regexp
+ * Returns list of matching files
+ * @param regexpString 
+ * @param searchTargets 
+ * @returns 
+ */
+function nodeGREP(regexpString: string, searchTargets: Array<string>): Array<string> {
+    let validSearchTargets: Array<string> = searchTargets.filter(target => fs.existsSync(target));
+    let invalidSearchTargets: Array<string> = searchTargets.filter(target => !fs.existsSync(target));
+
+    for (let target of invalidSearchTargets) {
+        console.log(`Target does not exist: ${target}`)
+    }
+
+    const regexp = new RegExp(regexpString);
+    const matchingFiles: Array<string> = [];
+
+    const fileMatchesRegExp = (filePath: string) => regexp.test(fs.readFileSync(filePath, 'utf8'));
```

Comment:
These little lamdas are really confusing. You should avoid them when they don't have to close over a variable in the scope, and when you can't avoid them you should give the more space to make it apparent what they do. It's better to pull them out into their own functions.